src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,203 @@
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.Rename;
+using Microsoft.VisualStudio.ProjectSystem.Waiting;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Rename
+{
+    [Order(Order.Default)]
+    [Export(typeof(IProjectTreeActionHandler))]
+    [Export(typeof(IFileRenameHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService)]
+    internal class RenamerProjectTreeActionHandler : ProjectTreeActionHandlerBase, IFileRenameHandler
+    {
+        private readonly IEnvironmentOptions _environmentOptions;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IVsUIService<IVsExtensibility, IVsExtensibility3> _extensibility;
+        private readonly IVsOnlineServices _vsOnlineServices;
+        private readonly IUserNotificationServices _userNotificationServices;
+        private readonly IWaitIndicator _waitService;
+        private readonly IRoslynServices _roslynServices;
+        private readonly Workspace _workspace;
+        private bool _isLanguageServiceDone;
+        private string _oldFilePath;
+
+        [ImportingConstructor]
+        public RenamerProjectTreeActionHandler(
+            UnconfiguredProject unconfiguredProject,
+            IUnconfiguredProjectVsServices projectVsServices,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace,
+            IEnvironmentOptions environmentOptions,
+            IUserNotificationServices userNotificationServices,
+            IRoslynServices roslynServices,
+            IWaitIndicator waitService,
+            IVsOnlineServices vsOnlineServices,
+            IProjectThreadingService threadingService,
+            IVsUIService<IVsExtensibility, IVsExtensibility3> extensibility)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _projectVsServices = projectVsServices;
+            _workspace = workspace;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _roslynServices = roslynServices;
+            _waitService = waitService;
+            _vsOnlineServices = vsOnlineServices;
+            _threadingService = threadingService;
+            _extensibility = extensibility;
+        }
+
+        public void HandleRename(string oldFilePath, string newFilePath)
+        {
+            if (_oldFilePath == oldFilePath)
+            {
+                _isLanguageServiceDone = true;
+            }
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            Assumes.Present(context);
+            Assumes.Present(node);
+            Assumes.Present(value);
```

Comment:
Use `Assumes.NotNull` here. `Assumes.Present` is for verifying services were retrieved successfully from service locators.

Reply:
Done (next commit).